docs/proposals/acting-as.md

```diff
@@ -0,0 +1,219 @@
+  old way for several release cycles.
+- change apiserver to recognize an "Acting-As" header in requests, and to check whether the
+  originally authenticated user is able to act as that user.
+- Record both originally authenticated user and acting-as user in the apiserver logs for
+  auditability. (Bonus).
+- change controller-manager to set "Acting-As" header in requests.  It will be hardcoded to use
+  the default service account of that namespace when making a request.  
+
+### Controllers other than replication controller
+
+- Job manager: should change the same as replication controller
+- DaemonSet manager: should change the same as replication controller
+- Endpoint controller: not sure if needed to use "Acting-As".
+- garbage collector: not sure if needed to use "Acting-As".
+- node controller: not sure if needed to use "Acting-As".
+- route controller: not sure if needed to use "Acting-As".
+- deployment controller: should use "Acting-As" when CRUDing replication controllers. 
+- service controller: does not create other object, so should not need to use "Acting-As".
+- quota controller: not sure if needed to use "Acting-As".
+- namespace controller: if it is executing initializers and finalizers, then it may need to "Acting-As" for some of those steps.  Needs thought.
+- horizontal-scaler controller: should use "Acting-As" to ensure it does not scale things the HPA creator should not be able to scale.
+- persistent-volume controller: not sure if needed to use "Acting-As".
+- persistent-volume recycler: not sure if needed to use "Acting-As"
+- serviceaccount.TokensController: probably should, since it CRUDs objects in response to changed made in other objects by users.
+- serviceaccount.ServiceAccountController: probably should, since it CRUDs objects in response to changed made in other objects by users.
+
+These do not all need to be convered in one go.  It should be possible to have an incremental transition.
+
+## Alternatives considered
+
+### Alternative: Access Review API
+
+In this alternate design, Controllers would not use Acting-As header, and they would not have
+authorization checks compiled into them.  Instead, they use an "Access Review" interface on the
+APIserver that lets a controller ask "If I were user A would I be allowed to create this Pod?".
+
+Problem with this is that a system admin has to fully trust any new controller (e.g. contributed
+controller).
+
+Recommendation: reject this alternative.
+
+### Alternative: Special-purpose accounts
+
+In this alternate design, the controller-manager authenticates
+as a special-purpose account called "controller-manager-acting-in-namespace-foo".
+No "acting-as" header.  Controller-manager actually stores credentials for all
+the accounts it is allowed to act as, and when a namespace is created,
+either the namespace admin consents to let controller-manager act in it,
+or a namespace initializer process creates the account.
```

Comment:
This is closer to the approach in OpenShift, with per-controller service accounts with (more) scoped permissions. We haven't reached per-controller-per-namespace segmentation, though.